### PR TITLE
Feature/switch from nginx to traefik

### PR DESCRIPTION
**Pull request by burnedikt**
- supersedes https://github.com/Innoactive/ansible-innoactive-hub-role/pull/68
- already in use for deployments to dev.hub.innoactive.de
- still missing support for the `letsencrypt` flag

### CODE REVIEW

**Comment by burnedikt** [comment]
molecule tests are passing, too 💪

File: defaults/main.yml [hunk 1]
```diff
@@ -61,6 +57,7 @@ hub_configuration:
   # hostname can contain multiple hostnames but we only need one
   # for most following options
   primary_hostname: "{{ hostname.split(',')[0] | trim }}"
+  hostname_aliases: "{{ hostname.split(',')[1:] | trim }}"
```
Comment (LukasReindlGit) [inline defaults/main.yml:60]: I think this throws an error when there is only one hostname

File: templates/python/enable_micro_frontend.py.j2 [hunk 2]
```diff
@@ -1,17 +1,20 @@
...
-# create entry for customization micro frontend
-customization_frontend, created = HubService.objects.get_or_create(name='{{ micro_frontend.name }}',
-                                                                   menu_entry_name='{{ micro_frontend.menu_entry_name }}',
-                                                                   url_prefix='{{ micro_frontend.url_prefix }}',
-                                                                   fa_icon='{{ micro_frontend.fa_icon }}',
-                                                                   script_url='{{ micro_frontend.script_url }}',
-                                                                   init_function='{{ micro_frontend.init_function }}')
+if HubService.objects.filter(name='Discovery Portal Customization').exists() is False:
+    # create entry for customization micro frontend
+    customization_frontend, created = HubService.objects.get_or_create(name='{{ micro_frontend.name }}',
```
Comment (burnedikt) [inline templates/python/enable_micro_frontend.py.j2:8]:
```suggestion
    customization_frontend, created = HubService.objects.create(name='{{ micro_frontend.name }}',
```

File: tasks/reverse_proxy.yml [hunk 3]
```diff
@@ -33,8 +33,8 @@
     env: "{{ env_vars | combine(letsencrypt_staging_env_vars) }}"
     labels:
       # Define a middleware that can redirect from http to https
-      traefik.http.middlewares.https-redirect.redirectscheme.scheme: "https"
-      # traefik.http.middlewares.https-redirect.redirectscheme.permanent: "true"
+      # traefik.http.middlewares.https-redirect.redirectscheme.scheme: "https"
+      traefik.http.middlewares.redirect-https.redirectscheme.scheme: https
```
Comment (burnedikt) [inline tasks/reverse_proxy.yml:37]:
```suggestion
      traefik.http.middlewares.redirect-https.redirectscheme.scheme: https
```

File: tasks/main_services.yml [hunk 4]
```diff
@@ -16,7 +16,7 @@
       DJANGO_ALLOWED_HOSTS: "{{ hub_configuration.hostname | mandatory }}"
       DJANGO_SECRET_KEY: "{{ hub_configuration.secret_key | mandatory }}"
       FROM_EMAIL: "{{ hub_configuration.from_email | default(hub_configuration.admin_email, true) | mandatory }}"
-      RAVEN_DSN: "{{ hub_configuration.sentry_dsn | default('') }}"
+      # RAVEN_DSN: "{{ hub_configuration.sentry_dsn | default('') }}"
```
Comment (burnedikt) [inline tasks/main_services.yml:19]: don't commit this.

File: tasks/discovery_portal.yml [hunk 5]
```diff
@@ -54,7 +54,7 @@
       traefik.http.middlewares.redirect-alternative-portal.redirectregex.regex: (.*\\/\\/)([^:\\/\\s]+)(.*)
       traefik.http.middlewares.redirect-alternative-portal.redirectregex.replacement: https://{{ portal_hostname }}$${3}
       traefik.http.routers.portal-alternative.rule: Host({{ traefik_alternative_hostnames }}) && Method(`GET`,`HEAD`)
-      traefik.http.routers.portal-alternative.middlewares: redirect-alternative-portal
+      traefik.http.routers.portal-alternative.middlewares: redirect-alternative-portal #,https-redirect
```
Comment (burnedikt) [inline tasks/discovery_portal.yml:57]: don't commit uncommented stuff

File: tasks/customization_service.yml [hunk 6]
```diff
@@ -83,6 +83,11 @@
       traefik.http.routers.customization.rule: Host(`{{ customization_hostname }}`)
       traefik.http.routers.customization.tls: "{{ letsencrypt | ternary('true', 'false') }}"
       traefik.http.routers.customization.tls.certresolver: lets-encrypt
+      # http redirect to https
+      traefik.http.routers.customization-http.rule: Host(`{{ customization_hostname }}`)
+      traefik.http.routers.customization-http.middlewares: redirect-https
+      traefik.http.routers.customization-http.entrypoints: http
+      traefik.http.routers.customization-http.priority: '999'
```
Comment (burnedikt) [inline tasks/customization_service.yml:90]: we only need this if `letsencrypt` is true.

File: tasks/discovery_portal.yml [hunk 7]
```diff
@@ -77,6 +77,13 @@
       traefik.http.routers.portal.rule: Host(`{{ portal_hostname }}`)
       traefik.http.routers.portal.tls: "{{ letsencrypt | ternary('true', 'false') }}"
       traefik.http.routers.portal.tls.certresolver: lets-encrypt
+
+      # http redirect to https
+      traefik.http.routers.portal-http.rule: Host(`{{ portal_hostname }}`)
+      traefik.http.routers.portal-http.middlewares: redirect-https
+      traefik.http.routers.portal-http.entrypoints: http
+      traefik.http.routers.portal-http.priority: '999'
```
Comment (burnedikt) [inline tasks/discovery_portal.yml:85]: we only need this if `letsencrypt` is true.

File: tasks/main_services.yml [hunk 8]
```diff
@@ -74,6 +74,12 @@
       traefik.http.routers.web.tls: "{{ letsencrypt | ternary('true', 'false') }}"
       traefik.http.routers.web.tls.certresolver: lets-encrypt
 
+      # http redirect to https
+      traefik.http.routers.web-http.rule: Host(`{{ hub_configuration.primary_hostname }}`)
+      traefik.http.routers.web-http.middlewares: redirect-https
+      traefik.http.routers.web-http.entrypoints: http
+      traefik.http.routers.web-http.priority: '999'
```
Comment (burnedikt) [inline tasks/main_services.yml:81]: we only need this if `letsencrypt` is true.

**Comment by burnedikt** [comment]
Open Topic:

- fluentd with new hostname?

File: defaults/main.yml [hunk 9]
```diff
@@ -10,7 +10,8 @@ docker_registry:
   password: "{{ registry_password }}"
 
 # Main Hub Configuration parameters
-hostname:
```
Comment (burnedikt) [inline defaults/main.yml:13]: breaking change

File: defaults/main.yml [hunk 10]
```diff
@@ -10,7 +10,8 @@ docker_registry:
...
+admin_hostname:
+admin_alias_hostnames: []
```
Comment (burnedikt) [inline defaults/main.yml:14]: both params need documentation in readme.md

File: defaults/main.yml [hunk 11]
```diff
@@ -22,25 +23,27 @@ setup_discovery_portal: true
 setup_customization_service: true
 hub_offering: "lean"
 concurrent_access_tokens: true
+traefik_dashboard: false
```
Comment (burnedikt) [inline defaults/main.yml:26]: docs missing

File: defaults/main.yml [hunk 12]
```diff
@@ -22,25 +23,27 @@ setup_discovery_portal: true
...
 # Error Tracking configuration
 sentry_dsn:
 # Google Analytics configuration
 google_analytics_tracking_id:
 extra_environment_variables: {}
 
+# Fluentd configuration
+fluentd_hostname:
```
Comment (burnedikt) [inline defaults/main.yml:34]: docs missing

File: defaults/main.yml [hunk 13]
```diff
@@ -22,25 +23,27 @@ setup_discovery_portal: true
...
+reverse_proxy_image_version: 2.2
 
 mailserver_image: namshi/smtp:latest
 message_queue_image: redis:5
 database_image: postgres:9.6
 hub_image: registry.docker.innoactive.de/hub_web:{{ hub_image_version }}
 portal_image: registry.docker.innoactive.de/hub_portal_web:{{ portal_image_version }}
-reverse_proxy_image: registry.docker.innoactive.de/hub_nginx:{{ reverse_proxy_image_version }}
-reverse_proxy_config_generator_image: jwilder/docker-gen:latest
-letsencrypt_proxy_companion_image: jrcs/letsencrypt-nginx-proxy-companion:v1.12
```
Comment (burnedikt) [inline defaults/main.yml:43]: breaking change

File: defaults/main.yml [hunk 14]
```diff
@@ -78,7 +81,8 @@ hub_configuration:
   extra_environment_variables: "{{ extra_environment_variables }}"
 
 # Portal Configuration
-portal_hostname: "portal.{{ hub_configuration.primary_hostname }}"
+portal_hostname:
+portal_alias_hostnames: []
```
Comment (burnedikt) [inline defaults/main.yml:85]: docs

File: defaults/main.yml [hunk 15]
```diff
@@ -102,13 +107,15 @@ launcher_configuration:
   oauth2_client_secret: "{{ launcher_oauth_client_secret }}"
 
 # Customization Configuration
-customization_hostname: "customization.{{ hub_configuration.primary_hostname }}"
+customization_hostname:
```
Comment (burnedikt) [inline defaults/main.yml:110]: breaking change ( no more default value), i.e. variable is now mandatory

File: defaults/main.yml [hunk 16]
```diff
@@ -102,13 +107,15 @@ launcher_configuration:
...
+customization_alias_hostnames: []
```
Comment (burnedikt) [inline defaults/main.yml:111]: docs

File: defaults/main.yml [hunk 17]
```diff
@@ -78,7 +81,8 @@ hub_configuration:
   extra_environment_variables: "{{ extra_environment_variables }}"
 
 # Portal Configuration
-portal_hostname: "portal.{{ hub_configuration.primary_hostname }}"
+portal_hostname:
```
Comment (burnedikt) [inline defaults/main.yml:84]: breaking change (no more default), i.e. variable now mandatory

**Pull request merged**